File: backend/app/schemas/credits.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from backend.app.models.types import UserRole

FreeCreditRequestStatus = Literal["pending", "approved", "rejected", "manual_review_required", "expired"]
FreeCreditSource = Literal["general", "linkedin_connection", "beta_tester", "manual_admin"]
PaymentAttemptStatus = Literal[
    "pending_payment",
    "temporarily_approved",
    "confirmed_paid",
    "rejected_payment",
    "expired_temporary_access",
    "manual_review_required",
]


class AdminGrantCreditRequest(BaseModel):
    user_id: str = Field(min_length=1)
    credit_amount_usd: float = Field(gt=0)
    note: str | None = Field(default=None, max_length=500)


class AdminGrantCreditResponse(BaseModel):
    user_id: str
    current_credit_usd: float


class AdminSetUserActiveRequest(BaseModel):
    user_id: str = Field(min_length=1)
    active: bool


class AdminSetUserActiveResponse(BaseModel):
    user_id: str
    is_active: bool
    current_credit_usd: float
    credit_floor_usd: float


class AdminUserSummary(BaseModel):
    id: str
    email: str | None
    full_name: str | None
    image_url: str | None = None
    role: UserRole
    is_active: bool
    current_credit_usd: float
    credit_floor_usd: float
    created_at_ms: int
    last_sign_in_at_ms: int | None


class AdminUserListResponse(BaseModel):
    items: list[AdminUserSummary]
    limit: int
    offset: int
    has_more: bool
    query: str | None = None


class PaymentIntegrationResponse(BaseModel):
    provider: str
    checkout_enabled: bool
    receipt_upload_enabled: bool
    reason: str | None = None
    paypal_recipient_email: str | None = None
    paypal_payment_url: str | None = None
    min_payment_usd: float
    max_payment_usd: float


class PayPalPaymentAttemptCreateRequest(BaseModel):
    expected_amount_usd: float = Field(gt=0)


class PaymentAttemptSummary(BaseModel):
    id: str
    user_id: str
    provider: str
    expected_amount_usd: float
    expected_currency: str
    reference_code: str
    status: PaymentAttemptStatus
    temporary_access_expires_at: datetime | None = None
    provider_reference: str | None = None
    credit_grant_id: str | None = None
    receipt_filename: str | None = None
    review_reason: str | None = None
    decision_note: str | None = None
    created_at: datetime
    updated_at: datetime


class PaymentAttemptListResponse(BaseModel):
    attempts: list[PaymentAttemptSummary] = Field(default_factory=list)


class AdminPaymentAttemptDecisionRequest(BaseModel):
    attempt_id: str = Field(min_length=1)
    status: Literal["confirmed_paid", "rejected_payment", "manual_review_required"]
    decision_note: str = Field(min_length=1, max_length=500)
    credit_amount_usd: float | None = Field(default=None, gt=0)
    provider_reference: str | None = Field(default=None, max_length=255)


class FreeCreditRequestCreate(BaseModel):
    requested_amount_usd: float | None = Field(default=None, gt=0)
    source: FreeCreditSource = "general"
    reason: str = Field(min_length=1, max_length=1000)
    linkedin_profile_url: str | None = Field(default=None, max_length=2048)
    relationship_note: str | None = Field(default=None, max_length=1000)
    intended_use: str | None = Field(default=None, max_length=1000)
    idempotency_key: str | None = Field(default=None, max_length=255)


class FreeCreditRequestSummary(BaseModel):
    id: str
    user_id: str
    requested_amount_usd: float | None = None
    source: FreeCreditSource
    reason: str
    linkedin_profile_url: str | None = None
    relationship_note: str | None = None
    intended_use: str | None = None
    evidence_verified: bool = False
    idempotency_key: str | None = None
    status: FreeCreditRequestStatus
    decided_amount_usd: float | None = None
    decision_note: str | None = None
    reviewer_user_id: str | None = None
    credit_grant_id: str | None = None
    created_at: datetime
    updated_at: datetime
    decided_at: datetime | None = None


class FreeCreditRequestListResponse(BaseModel):
    requests: list[FreeCreditRequestSummary] = Field(default_factory=list)


class AdminFreeCreditDecisionRequest(BaseModel):
    request_id: str = Field(min_length=1)
    status: Literal["approved", "rejected", "manual_review_required"]
    credit_amount_usd: float | None = Field(default=None, gt=0)
    decision_note: str = Field(min_length=1, max_length=500)

File: frontend/src/types/credits.ts
import type {
  CreditGrantRecord,
  FreeCreditRequestRecord,
  FreeCreditRequestStatus,
  PaymentAttemptRecord,
  PaymentAttemptStatus,
  ManualCreditGrantRequest,
  UserRole,
} from "../../../vendor/ai-portfolio-admin/frontend/types";

export type { FreeCreditRequestStatus, PaymentAttemptStatus };

export type GrantCreditPayload = Pick<ManualCreditGrantRequest, "user_id" | "credit_amount_usd"> & {
  note?: string;
};

export type GrantCreditResponse = Pick<CreditGrantRecord, "user_id"> & {
  current_credit_usd: number;
};

export type SetUserActivePayload = {
  user_id: string;
  active: boolean;
};

export type SetUserActiveResponse = {
  user_id: string;
  is_active: boolean;
  current_credit_usd: number;
  credit_floor_usd: number;
};

export type AdminUserSummary = {
  id: string;
  email: string | null;
  full_name: string | null;
  image_url?: string | null;
  role: UserRole;
  is_active: boolean;
  current_credit_usd: number;
  credit_floor_usd: number;
  created_at_ms: number;
  last_sign_in_at_ms: number | null;
};

export type AdminUserListResponse = {
  items: AdminUserSummary[];
  limit: number;
  offset: number;
  has_more: boolean;
  query: string | null;
};

export type PaymentIntegrationResponse = {
  provider: string;
  checkout_enabled: boolean;
  receipt_upload_enabled: boolean;
  reason: string | null;
  paypal_recipient_email: string | null;
  paypal_payment_url: string | null;
  min_payment_usd: number;
  max_payment_usd: number;
};

export type PayPalPaymentAttemptCreateRequest = {
  expected_amount_usd: number;
};

export type PaymentAttemptSummary = Omit<PaymentAttemptRecord, "user_id" | "created_at"> & {
  user_id: string;
  credit_grant_id: string | null;
  receipt_filename: string | null;
  review_reason: string | null;
  decision_note: string | null;
  created_at: string;
  updated_at: string;
};

export type PaymentAttemptListResponse = {
  attempts: PaymentAttemptSummary[];
};

export type AdminPaymentAttemptDecisionRequest = {
  attempt_id: string;
  status: Extract<PaymentAttemptStatus, "confirmed_paid" | "rejected_payment" | "manual_review_required">;
  decision_note: string;
  credit_amount_usd?: number | null;
  provider_reference?: string | null;
};

export type FreeCreditRequestCreate = {
  requested_amount_usd?: number | null;
  source?: "general" | "linkedin_connection" | "beta_tester" | "manual_admin";
  reason: string;
  linkedin_profile_url?: string | null;
  relationship_note?: string | null;
  intended_use?: string | null;
  idempotency_key?: string | null;
};

export type FreeCreditRequestSummary = FreeCreditRequestRecord & {
  updated_at: string;
};

export type FreeCreditRequestListResponse = {
  requests: FreeCreditRequestSummary[];
};

export type AdminFreeCreditDecisionRequest = {
  request_id: string;
  status: Extract<FreeCreditRequestStatus, "approved" | "rejected" | "manual_review_required">;
  credit_amount_usd?: number | null;
  decision_note: string;
};
